ngcc - handle multiple basePaths & use typings for dependency resolution

petebacondarwin opened the pull request:
Resolves FW-1204

alan-agius4 commented on packages/compiler-cli/ngcc/src/packages/entry_point_finder.ts:
```diff
@@ -21,15 +23,52 @@ export class EntryPointFinder {
+        basePaths.push(AbsoluteFsPath.fromUnchecked(join(baseUrl, extractPathPrefix(path))));
+      }));
+    }
+    basePaths.sort();  // Get the paths in order with the shorter ones first.
```
Changing the order might break the resolutions.

Example, when having some like the below
```
paths: {
	"foo/*": [
		"./local-path/foo/*",
	],
        "*": [
              "node_modules/*"
         ]
}
```

First it should resolve in `local-path` fails it will fallback to node_modules, but with this logic it will be inverted.

The order of paths, in `paths` is quite critical.

petebacondarwin replied:
The aim of this code is to find all the potential entry points. It is independent of dependency resolution.
We want to find the minimal set of paths that cover all paths that could contain entry points.

petebacondarwin replied:
To be more concrete (assuming that `baseUrl` is `/my-project`).

In the 1st example:

```
paths: {
	"foo/*": [ "./local-path/foo/*"],
        "*": ["node_modules/*"]
}
```

The `basePaths` that we want to parse are `myProject/local-path/foo` and `myProject/node_modules`.

(Actually in this case there is no ordering here anyway. I don't believe that the order of the keys could be relied upon, and there is only one item in each array.)

In the 2nd example:

```
paths: {
	"foo/*/bar": ["./local-path/foo/*/bar"],
        "*": ["node_modules/*"]
}
```

The `basePaths` would again be `/myProject/local-path/foo` and `myProject/node_modules`.

In both cases, we are not interested (at this point in the processing) of how dependencies are resolved. We are just interested in getting some paths that we should walk to look for potential entry-points to process.

Later we will use the `paths` mappings "properly" to actually resolve the dependencies between entry-points.

petebacondarwin commented:
Re-opening with a new approach.

alxhub commented on packages/compiler-cli/ngcc/src/packages/module_resolver.ts:
```diff
@@ -0,0 +1,264 @@
+   * or the folder containing the `package.json` of the entry-point
+   * if not relative and not deep.
+   * */
+  resolvedPath: AbsoluteFsPath;
```
I would prefer we not mix two separate concepts into the same interface. Let's make `ResolvedModule` a union type between a `RelativeModule` and `AbsoluteModule` interface, with appropriately named fields for each? `isRelative` could be used to switch between them.

petebacondarwin replied:
Excellent idea!

alxhub commented on packages/compiler-cli/ngcc/src/packages/module_resolver.ts:
```diff
@@ -0,0 +1,264 @@
+}
+
+interface PathMappingPattern {
+  prefix: string, postfix: string, hasWildcard: boolean;
```
`;` not `,` for field separation, then the auto-formatter will lay it out properly.

alxhub commented on packages/compiler-cli/ngcc/src/packages/module_resolver.ts:
```diff
@@ -0,0 +1,264 @@
+   * `template.prefix` and `template.postfix` strings.
+   */
+  private findMappedPaths(moduleName: string): AbsoluteFsPath[] {
+    if (this.pathMappings.length > 0) {
```
Rather than guard the whole logic, instead introduce an early exit if there are no `pathMappings`.

petebacondarwin replied:
OK

petebacondarwin replied:
Actually I can avoid calling this method if `pathMappings.length === 0`.

alxhub commented on packages/compiler-cli/ngcc/src/packages/module_resolver.ts:
```diff
@@ -0,0 +1,264 @@
+  private findPackagePath(path: AbsoluteFsPath): AbsoluteFsPath|null {
+    let folder = path;
+    while (folder !== '/') {
+      folder = AbsoluteFsPath.fromUnchecked(dirname(folder));
```
This operation happens a couple times here. We could consider introducing `AbsoluteFsPath.dirname(folder)` as an API which performs this operation. It would also help get away from having to import the node API `path` everywhere.

petebacondarwin replied:
Happy to do so if you are comfortable with adding stuff to that library.

alxhub commented on packages/compiler-cli/ngcc/src/packages/module_resolver.ts:
```diff
@@ -0,0 +1,264 @@
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import * as fs from 'fs';
```
How hard do you think it would be to make `ModuleResolver` depend on `ts.ModuleResolutionHost` instead?

That would allow the tests to not have to constantly mock/restore `fs` and simply provide a mock filesystem.

petebacondarwin replied:
Actually I would rather implement our own `FileSystemHost`, similar to what `enhanced-resolve` does. This would also allow us to do caching, as well as mocking, which would speed up module resolution. See https://github.com/webpack/enhanced-resolve/blob/master/lib/node.js#L12-L15

petebacondarwin replied:
Or maybe use the one from `enhanced-resolve`??

petebacondarwin replied:
We could also then use such as general file host inside a `CompilerHost`, which is a `ModuleResolutionHost` and could give us that caching opportunities.

petebacondarwin replied:
I have implemented a `FileSystem` interface.

alxhub commented on packages/compiler-cli/ngcc/test/packages/dependency_host_spec.ts:
```diff
@@ -81,125 +64,113 @@ describe('DependencyHost', () => {
-        '/internal/inner.js': `import {Y} from 'path/to/y';`,
-        '/internal/circular-a.js': `import {B} from './circular-b'; import {X} from 'path/to/x';`,
-        '/internal/circular-b.js': `import {A} from './circular-a'; import {Y} from 'path/to/y';`,
+        '/no/imports/or/re-exports/index.js': '// some text but no import-like statements',
```
I appreciate the convenience, but I'm not a fan of test suites which have one single large "data set" that contains the mock data for each test.

1) It can introduce accidental cross-dependencies between tests, where something present in the data for one test causes another to behave differently. This has already caused issues for View Engine tests written in this style when porting to Ivy.
2) It's harder to follow the individual tests as the data is elsewhere in the file.

Even though it's more verbose, can we refactor this so each test sets up only the test files for its specific purpose?

petebacondarwin replied:
Actually one of the reasons I have not broken this up is that havng those potential cross-dependencies helps to ensure that we are not overly simplifying each unit test. In other words, it is easy to make the unit test too narrow and not to cover potential corner cases.

But I do agree about the second point. I'll see what I can do to make things better.

petebacondarwin replied:
I have now implemented the `MockFileSystem` class but there are already many many changes in this PR. And there are lots of tests that use large shared data-sets.

I will follow up after the Easter holidays with a PR to break up the data-sets, where appropriate.

petebacondarwin replied:
See FW-1252

benlesh commented:
Sorry, @petebacondarwin ... I readded the "review" label, because I'm just trying to track what needs reviewed and not while caretaking. :heart:

AndrewKushnir commented:
@petebacondarwin could you please confirm that this change should go to master-only since RC is out? Thank you.

petebacondarwin committed:
refactor(ivy): ngcc - tidy up `DependencyResolver` helper method

This method was poorly named for what it does, and did not have a
return type.

petebacondarwin committed:
test(ivy): ngcc - check private dependency in integration test

The test now attempts to compile an entry-point (@angular/common/http/testing)
that has a transient "private" dependency. A private dependency is one that is
only visible by looking at the compiled JS code, rather than the generated TS
typings files.

This proves that we can't rely on typings files alone for computing the
dependencies between entry-points.

petebacondarwin committed:
refactor(compiler-cli): ngcc - track non-Angular entry-points

Previously we completely ignored entry-points that had not been
compiled with Angular, since we do not need to compile them
with ngcc. But this makes it difficult to reason about dependencies
between entry-points that were compiled with Angular and those that
were not.

Now we do track these non-Angular compiled entry-points but they
are marked as `compiledByAngular: false`.

petebacondarwin committed:
refactor(compiler-cli): ngcc - remove unnecessary `sourcePath` parameters

The `Transformer` and `Renderer` classes do not
actually need a `sourcePath` value as by the time
they are doing their work we are only working directly
with full absolute paths.

petebacondarwin committed:
refactor(ivy): ngcc - tidy up `mainNgcc`

petebacondarwin committed:
feat(ivy): add helper methods to AbsoluteFsPath

petebacondarwin committed:
refactor(ivy): ngcc - implement abstract FileSystem

This commit introduces a new interface, which abstracts access
to the underlying `FileSystem`. There is initially one concrete
implementation, `NodeJsFileSystem`, which is simply wrapping the
`fs` library of NodeJs.

Going forward, we can provide a `MockFileSystem` for test, which
should allow us to stop using `mock-fs` for most of the unit tests.
We could also implement a `CachedFileSystem` that may improve the
performance of ngcc.

petebacondarwin committed:
refactor(ivy): ngcc - implement new module resolver

When working out the dependencies between entry-points
ngcc must parse the import statements and then resolve the
import path to the actual file.  This is complicated because module
resolution is not trivial.

Previously ngcc used the node.js `require.resolve`, with some
hacking to resolve modules. This change refactors the `DependencyHost`
to use a new custom `ModuleResolver`, which is optimized for this use
case.

Moreover, because we are in full control of the resolution,
we can support TS `paths` aliases, where not all imports come from
`node_modules`. This is the case in some CLI projects where there are
compiled libraries that are stored locally in a `dist` folder.
See //FW-1210.

petebacondarwin committed:
test(ivy): ngcc - tighten up typings in Esm5ReflectionHost specs

petebacondarwin committed:
refactor(ivy): ngcc - add MockFileSystem

petebacondarwin committed:
feat(ivy): ngcc - support additional paths to process

By passing a `pathMappings` configuration (a subset of the
`ts.CompilerOptions` interface), we can instuct ngcc to process
additional paths outside the `node_modules` folder.

petebacondarwin committed:
refactor(ivy): ngcc - move the dependency resolving stuff around

For UMD/RequireJS support we will need to have multiple
`DependencyHost` implementations. This commit  prepares the
ground for that.

petebacondarwin committed:
refactor(ivy): ngcc - remove the last remnants of `path` and `canonical-path`

The ngcc code now uses `AbsoluteFsPath` and `PathSegment` to do all its
path manipulation.

petebacondarwin commented:
Just merging into master now that RC has been released.
